### ISSUE TITLE
Better unit testing support

### ISSUE THREAD

**Issue by paulmillr** [brunch/brunch#988]
Here's a quote from a long-time brunch user

> I have been using Brunch for projects for the last 4 years. It actually has been a fantastic experience because once you build your own skeleton, you can easily reuse it for other projects. My biggest concern with it was the fact that it hijacks require making unit tests very difficult. You end up having to write unit tests on your compiled package rather than the individual modules. Some people prefer this, but for me, working with node over the last 5 years, I have adopted the modular approach and enjoy writing a small file and writing a test for it.
> 
> I ended up trying Grunt and Gulp and found that Grunt sits nicer with me. I was able to get pretty much all your expectations working, and once you get your skeleton set up, just like I did with Brunch, I can re use it for multiple different projects. I am still tweaking my Grunt build system to make it perfect for what I need, but so far, I am happy with the switch because now I can unit test my code.
> 
> TL;DR; Started with Brunch 4 years ago. Switched to Grunt and love it. Can unit test my code which is a must.

**Comment by nblackburn** [comment]
I think this issue needs more attention as it seems to be a fairly large hinderance.

**Comment by goshakkk** [comment]
I think we (as in the brunch community, not necessarily the maintainers) need to identify the current pain points of unit testing with brunch.

If the `require` is the only issue, then it seems totally workable. It seems in the land of webpack, officially, unit testing means either not using enhanced require functionality, or compiling (https://github.com/webpack/docs/wiki/testing).

Though I've also recently seen some people do something more interesting — not using webpack at all when it comes to unit tests. It basically relies on mocking `require` (https://medium.com/@TomazZaman/how-to-get-fast-unit-tests-with-out-webpack-793c408a076f)

Which is an interesting technique, and I believe something like that could also be done with Brunch, assuming you author in JS (or CoffeeScript and `coffeescript.register()` is called) — though not sure how that could work out for other source languages.

The other alternative seems to be actually compiling. I think entry points could even be utilized to produce a "compiled" bundle for each test file.

Does anyone have any input on all of this? Willing to share experiences, and the ways you wish it worked instead?

**Comment by nblackburn** [comment]
> Though I've also recently seen some people do something more interesting — not using webpack at all when it comes to unit tests. It basically relies on mocking require (https://medium.com/@TomazZaman/how-to-get-fast-unit-tests-with-out-webpack-793c408a076f)

This is normally what i would do but you quickly hit a roadblock when you are testing something that isn't javascript (_like a view_) as you can't use **entryPoints** leaving no choice but to test on the compiled source of the entire application.

**Comment by goshakkk** [comment]
The current restrictions on limiting entryPoints to javascripts are more like a way to start with one case, but one that works well. Given that internally Brunch treats JS and template files the same way, it's technically possible to also allow entry-point'ing templates.

---

One idea I've just had is... could we build some kind of a communication "port" into the watcher, which would allow tests to tell the watcher which file they want `require`d, and the watcher will give a compiled version back... taking into account all the enhancements to `require` that Brunch provides. It's probably a little bit more involved on the implementation side of things, esp when it comes to customizing `require`, but yeah.,I think something like this could work.

**Comment by nblackburn** [comment]
Allowing templates in entryPoints would be a major improvement and provide a means of testing isolated components.

To be honest, i think i need to open up a request for a babel-plugin for the framework i use as React has one so it's not really a problem for people using that.

**Comment by goshakkk** [comment]
The binding of require into the watcher though would be way more flexible and simpler to use, freeing the users of the need to create a separate entry point for every file they would want tot test in isolation. I will be looking into this

**Comment by nblackburn** [comment]
I am glad this is being looked at, it's a real pain point for me.

**Comment by jmls** [comment]
has there been any more thoughts or progress on this ? I've just stumbled into this problem as well ;)

**Comment by nblackburn** [comment]
@goshakkk Any updates on this?
